### ISSUE TITLE
Potential slow SQL query in Products Low In Stock `/products/low-in-stock` endpoint.

### ISSUE THREAD

**Issue by kaushikasomaiya** [woothemes/woocommerce#38932]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Opening this up for visibility and investigation. This report will help merchants facing similar issues.

6458996-zen

A new endpoint was introduced here for pulling Low In Stock products: https://github.com/woocommerce/woocommerce-admin/issues/7358#issuecomment-883908480 to resolve a similar slow query earlier.

This endpoint is called from some places like the **WooCommerce > Home > Stock** widget:

![https://d.pr/i/YStdZP](https://d.pr/i/YStdZP+) 
Full Size: https://d.pr/i/YStdZP

Merchant reports their site times out when store admins are working in the backend.

They have following specs

```
WC Version: 7.8.0
REST API Version: ✔ 7.8.0
WC Blocks Version: ✔ 10.2.3
Action Scheduler Version: ✔ 3.6.1
Log Directory Writable: ✔
WP Version: 6.2.2
WP Multisite: ✔
WP Memory Limit: 4 GB
WP Debug Mode: –
WP Cron: –
Language: nl_NL
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.57 (Debian)
PHP Version: 7.4.33
PHP Post Max Size: 250 MB
PHP Time Limit: 900
PHP Max Input Vars: 10000
cURL Version: 7.64.0
OpenSSL/1.1.1n

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.4.20-MariaDB-1:10.4.20+maria~buster-log
Max Upload Size: 1 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.8.0
WC Database Prefix: wp_
Totale database grootte: 987.19MB
Database gegevens grootte: 556.88MB
Database index grootte: 430.31MB
```

The slow query generates from 'products/low-in-stock' endpoint. Below is the slow query log shared by the merchant. 

```
SELECT
   SQL_CALC_FOUND_ROWS wp_posts.*,
   meta.meta_value AS low_stock_amount,
   wc_product_meta_lookup.stock_quantity 
FROM
   wp_wc_product_meta_lookup wc_product_meta_lookup 
   LEFT JOIN
      wp_posts wp_posts 
      ON wp_posts.ID = wc_product_meta_looku p.product_id 
   LEFT JOIN
      wp_postmeta AS meta 
      ON wp_posts.ID = meta.post_id 
      AND meta.meta_key = '_low_stock_amount' 
WHERE
   wp_posts.post_type IN 
   (
      'product',
      'product_variation'
   )
   AND wp_posts.post_status = 'publish' 
   AND wc_product_meta_lookup.st ock_quantity IS NOT NULL 
   AND wc_product_meta_lookup.stock_status IN
   (
      'instock',
      'outofstock'
   )
   AND 
   (
( meta.meta_value > '' 
      AND wc_product_meta_lookup.stock_quantity <= CAST( meta.meta_value AS SIGNED ) ) 
      OR 
      (
( meta.meta_value IS NULL 
         OR meta.meta_value <= '' ) 
         AND wc_product_meta_lookup.stock_quantity <= 1 
      )
   )
order by
   wc_product_meta_lookup.product_id DESC limit 0,
   1;

```

Above is a more expensive version of the query because "Low stock threshold" is configured on a per product basis.

https://github.com/woocommerce/woocommerce/blob/ba5cc19757459384361981c931ff9d4574345fc8/plugins/woocommerce/src/Admin/API/ProductsLowInStock.php#L249

I've opened this up so that we can track more instances and devs can investigate as necessary.

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by kaushikasomaiya** [comment]
To add:

I disabled stock management on my site and could still see the low-in-stock endpoint being called on WooCommerce > Home area:

![https://d.pr/i/6QbHP0](https://d.pr/i/6QbHP0+) 

Edit: This endpoint seems to be called on Product edit pages as well as other WooCommerce Admin pages even when Stock management is disabled.

**Comment by kaushikasomaiya** [comment]
One workaround I could think of is disabling the end-point if the Query is slowing down the server. 

```
add_filter('rest_endpoints', function( $endpoints ) {

    foreach( $endpoints as $route => $endpoint ){

			if (strpos($route, 'products/low-in-stock') !== false) {
						unset( $endpoints[ $route ] );
			}
    }

    return $endpoints;
});
```

**Comment by jojoba007** [comment]
To prove the issue lies in WooCommerce core: We cloned the WP MultiSite with the issue to a new server and made a screen recording of it. As you can see in the recording we disabled all network plugins and network themes. Also we disabled all plugins on individual site level, except for WooCommerce and Query Monitor to monitor the query issue. As you can see in the recording after disabling everything (except WooCommerce and Query Monitor) the issue is still being fired. In the video you can also see the issue popping up in our server query monitor. As soon as we try to edit a product, the slow stock query is being fired again and our server resources are being eaten and stressed to the maximum, resulting in 524 server errors.

See link to video: https://drive.google.com/file/d/1Q6tyYootQkB23izvuvYlXmjfOYzRBP1Q/view?usp=sharing

Download the videofile it if it wont play directly.

**Comment by moon0326** [comment]
Hey @woocommerce/mothra ✋ 

Would you mind if I assign this issue to myself? I'm already familiar with the issue as I worked on the [previous issue](https://github.com/woocommerce/woocommerce-admin/issues/7358#issuecomment-883908480). We have a HACK week,  and this could be a good issue to fix.

**Comment by louwie17** [comment]
> Would you mind if I assign this issue to myself? I'm already familiar with the issue as I worked on the https://github.com/woocommerce/woocommerce-admin/issues/7358#issuecomment-883908480. We have a HACK week, and this could be a good issue to fix.

Hey @moon0326, yes for sure, more then happy for you to take this :)

**Comment by moon0326** [comment]
> > Would you mind if I assign this issue to myself? I'm already familiar with the issue as I worked on the [woocommerce/woocommerce-admin#7358 (comment)](https://github.com/woocommerce/woocommerce-admin/issues/7358#issuecomment-883908480). We have a HACK week, and this could be a good issue to fix.
> 
> Hey @moon0326, yes for sure, more then happy for you to take this :)

Actually, Ghidraoh will be in Japan for a meet-up and I might not get a chance to work on it until a week after next week 🤦 . Please feel free to start it.

**Comment by moon0326** [comment]
Hey @AnnaMag ✋ 

I'm actually working on this. Do you mind if I take it over?

**Comment by moon0326** [comment]
Some updates:

I've tested the issue on Mariadb 10.4.20, 11.0.2 (latest), and Mysql 8.0.33. The query was slowest in MariaDB 11.0.2, but it did not cause a timeout issue (Avg. execution time was ~200ms).  All three had 82,001 rows in `wp_posts` and 51,908 rows in `wp_wc_product_meta_lookup` (21640 products). 

On MariaDB 11.0.2, I noticed it's not using the index as expected (see Using temporary; and Using filesort)

![Screen Shot 2023-08-02 at 7 28 21 PM](https://github.com/woocommerce/woocommerce/assets/4723145/14637999-104e-4105-b116-c15f52a7a278)

This can be easily fixed by changing `order by wc_product_meta_lookup.product_id desc` to `order by wp_posts.ID desc`

@jojoba007 Could you re-open your test server one more time, please? I thought the DB server version was the culprit, but it seems like it's not. I wasn't able to reproduce the issue on MariaDB 10.4.20 (This is what's on your test server). 


I'm going to put a hold on this issue for now until we can have a setup where we can constantly reproduce the issue.

**Comment by jojoba007** [comment]
> Some updates:
> 
> I've tested the issue on Mariadb 10.4.20, 11.0.2 (latest), and Mysql 8.0.33. The query was slowest in MariaDB 11.0.2, but it did not cause a timeout issue (Avg. execution time was ~200ms). All three had 82,001 rows in `wp_posts` and 51,908 rows in `wp_wc_product_meta_lookup` (21640 products).
> 
> On MariaDB 11.0.2, I noticed it's not using the index as expected (see Using temporary; and Using filesort)
> 
> ![Screen Shot 2023-08-02 at 7 28 21 PM](https://user-images.githubusercontent.com/4723145/257984070-14637999-104e-4105-b116-c15f52a7a278.jpg)
> 
> This can be easily fixed by changing `order by wc_product_meta_lookup.product_id desc` to `order by wp_posts.ID desc`
> 
> @jojoba007 Could you re-open your test server one more time, please? I thought the DB server version was the culprit, but it seems like it's not. I wasn't able to reproduce the issue on MariaDB 10.4.20 (This is what's on your test server).
> 
> I'm going to put a hold on this issue for now until we can have a setup where we can constantly reproduce the issue.

Hi @moon0326 I re-opened the test server. The credentials are still the same. Please let me know if you need additional info or credentials.

**Comment by moon0326** [comment]
> > Some updates:
> > I've tested the issue on Mariadb 10.4.20, 11.0.2 (latest), and Mysql 8.0.33. The query was slowest in MariaDB 11.0.2, but it did not cause a timeout issue (Avg. execution time was ~200ms). All three had 82,001 rows in `wp_posts` and 51,908 rows in `wp_wc_product_meta_lookup` (21640 products).
> > On MariaDB 11.0.2, I noticed it's not using the index as expected (see Using temporary; and Using filesort)
> > ![Screen Shot 2023-08-02 at 7 28 21 PM](https://user-images.githubusercontent.com/4723145/257984070-14637999-104e-4105-b116-c15f52a7a278.jpg)
> > This can be easily fixed by changing `order by wc_product_meta_lookup.product_id desc` to `order by wp_posts.ID desc`
> > @jojoba007 Could you re-open your test server one more time, please? I thought the DB server version was the culprit, but it seems like it's not. I wasn't able to reproduce the issue on MariaDB 10.4.20 (This is what's on your test server).
> > I'm going to put a hold on this issue for now until we can have a setup where we can constantly reproduce the issue.
> 
> Hi @moon0326 I re-opened the test server. The credentials are still the same. Please let me know if you need additional info or credentials.

Thank you so much 🙏

**Comment by kaushikasomaiya** [comment]
@moon0326 Just a note since you are working on it, this endpoint seems to be called from pages where it is not needed - for instance Products list page in WP Admin.

**Comment by moon0326** [comment]
@kaushikasomaiya Thank you. I'll look into it as well.

**Comment by moon0326** [comment]
@jojoba007 I've just checked the site and I'm getting 403 forbidden when I enter PHPMyAdmin. 
![Screen Shot 2023-08-07 at 11 47 53 AM](https://github.com/woocommerce/woocommerce/assets/4723145/cc99985b-d824-4773-ad6d-104d519ea632)

**Comment by jojoba007** [comment]
> @jojoba007 I've just checked the site and I'm getting 403 forbidden when I enter PHPMyAdmin.
> ![Screen Shot 2023-08-07 at 11 47 53 AM](https://user-images.githubusercontent.com/4723145/258900329-cc99985b-d824-4773-ad6d-104d519ea632.jpg)

Please, could you try again now @moon0326

**Comment by moon0326** [comment]
> > @jojoba007 I've just checked the site and I'm getting 403 forbidden when I enter PHPMyAdmin.
> > ![Screen Shot 2023-08-07 at 11 47 53 AM](https://user-images.githubusercontent.com/4723145/258900329-cc99985b-d824-4773-ad6d-104d519ea632.jpg)
> 
> Please, could you try again now @moon0326

I'm still getting the same error @jojoba007

**Comment by moon0326** [comment]
@jojoba007 In case you missed the last reply 🙏

**Comment by Khnaz35** [comment]
@moon0326 just sent you email, kindly have a look and let me know if you are still facing the same issue.

**Comment by moon0326** [comment]
Thanks @Khnaz35 👍 

@Khnaz35 sent me a sql dump via email for local testing. I was able to reproduce the issue on MariaDB 10.4.20, but not on the latest (11.0.2). 

Further investigation revealed that **SQL_CALC_FOUND_ROWS** is causing issues in MariaDB 10.4.x versions. Fortunately, it seems like this problem has been resolved in MariaDB 10.5.21+. I wasn't able to reproduce it in MariaDB 11.x.x versions.

I will continue to explore alternatives to determine the best course of action. It's likely that we will separate the **SQL_CALC_FOUND_ROWS** operation and execute it independently to enhance compatibility. As SQL_CALC_FOUND_ROWS has been deprecated since MySQL 8+, it's time to make a change IMO.


Note: there are many articles about **SQL_CALC_FOUND_ROWS**. You can start with https://www.percona.com/blog/to-sql_calc_found_rows-or-not-to-sql_calc_found_rows/
